Introduce `ActiveModel::Enum`

p8 opened the pull request:
This moves most of the implementation of ActiveRecord::Enum to
ActiveModel::Enum. This excludes:
* scopes, as scopes aren't supported by ActiveModel.
* method conflict detection, as this is difficult to implement in
  ActiveModel.

```ruby
class Conversation
  include ActiveModel::Attributes
  include ActiveModel::Enum
  attribute :status, :integer
  enum :status, [ :active, :archived ]
end

conversation = Conversation.new
conversation.active!
conversation.active? # => true
conversation.status  # => "active"
```

The current implementation supports using integers for values and
strings for labels, if defined.

This also introduces ActiveSupport::Concern to ActiveRecord::Enum as
this allows both ActiveModel::Enum and ActiveRecord::Enum to be mixed in
with `include` instead of `extend`, like most other ActiveModel modules.

The tests were copied from ActiveRecord and modified for ActiveModel.
Scoping and method conflict detection tests were excluded.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

adrianna-chang-shopify commented on activemodel/lib/active_model/enum.rb:
```diff
@@ -0,0 +1,347 @@
+  # are prefixed/suffixed with the name of the enum. It is also possible to
+  # supply a custom value:
+  #
+  #   class Conversation < ActiveRecord::Base
```
Should we avoid having classes inherit from `ActiveRecord::Base` in the docs?

p8 replied:
Yes. I think we should mostly avoid these in ActiveModel.

zzak commented:
cc @jonathanhefner and @ghiculescu who are probably interested in AM::Enum and would be good to get their feedback :pray:

jonathanhefner commented on activemodel/lib/active_model/attribute_registration.rb:
```diff
@@ -40,6 +40,10 @@ def attribute_types # :nodoc:
         end
       end
 
+      def type_for_attribute(attribute)
+        attribute_types[attribute.to_s]
+      end
```
Defining `type_for_attribute` like this here will shadow the [definition in `ActiveRecord::ModelSchema`](https://github.com/rails/rails/blob/d0f40f71b0b61f828291f366bad5d29ab44e70d2/activerecord/lib/active_record/model_schema.rb#L453-L462) because `ActiveRecord::Attributes` includes `ActiveModel::AttributeRegistration`, and it is included after `ActiveRecord::ModelSchema`.

I think it's worth splitting this out into its own PR, including tests, so I submitted #49910 and added you as a co-author.

p8 replied:
Nice! Thanks @jonathanhefner

jonathanhefner commented on activemodel/lib/active_model/attribute_methods/before_type_cast.rb:
```diff
@@ -1,10 +1,10 @@
-module ActiveRecord
+module ActiveModel
   module AttributeMethods
-    # = Active Record Attribute Methods Before Type Cast
+    # = Active Model Attribute Methods Before Type Cast
     #
-    # ActiveRecord::AttributeMethods::BeforeTypeCast provides a way to
+    # ActiveModel::AttributeMethods::BeforeTypeCast provides a way to
     # read the value of the attributes before typecasting and deserialization.
     #
     #   class Task < ActiveRecord::Base
```
`ActiveRecord::AttributeMethods::BeforeTypeCast` is a public module, so I don't think we can remove it without a deprecation.

I think this is also worth splitting out into a separate PR.  I actually made a while ago: #44534.  I've rebased it and added you as a co-author.

p8 replied:
Great! Thanks @jonathanhefner

p8 committed:
Use ActiveSupport::Concern for ActiveRecord::Enum

Allow ActiveRecord::Enum to be mixed in with include instead of
extend, like most other ActiveModel inherited modules.

This refactoring makes it easier to move `ActiveRecord::Enum` to
`ActiveModel::Enum` and mixin `ActiveModel::Enum` with include instead
of `extend`.

p8 committed:
Introduce ActiveModel::Enum

This moves most of the implementation of ActiveRecord::Enum to
ActiveModel::Enum. This excludes:

* scopes, as scopes aren't supported by ActiveModel.
* method conflict detection, as this is difficult to implement in
  ActiveModel.

```ruby
class Conversation
  include ActiveModel::Attributes
  include ActiveModel::Enum
  attribute :status, :integer
  enum :status, [ :active, :archived ]
end

conversation = Conversation.new
conversation.active!
conversation.active? # => true
conversation.status  # => "active"
```

The current implementation supports using integers for values and
strings for labels, if defined.

The tests were copied from ActiveRecord and modified for ActiveModel.
Scoping and method conflict detection tests were excluded.

rafaelfranca commented:
Thank you for the pull request. I didn't want Enum to be added to Active Record. For sure I'll not let it be added to Active Model. There is no reason to translate strings to integer in the active model layer. There is no benefits on doing that. I can see the benefits in Active Record.

p8 commented:
Hi @rafaelfranca,
You mentioned this at RailsWorld so I wasn't going to create this PR, but I saw Jonathan mentioning introducing ActiveModel::Enum in another PR so I thought I'd push my work 😄 

I don't care about the mapping of integers to strings either. Keeping that in place just made the PR easier/smaller.

My main reason for an ActiveModel::Enum is to replace code like the following:

```ruby
class SomeModel
  module States
    STARTED = 'started'
    DONE    = 'done'
    FAILED  = 'failed'
    ALL     = constants(false).map { |constant| const_get(constant) }
  end
  attribute :state, String
  validates :state, inclusion: { in: States::ALL }

  def started?
    state == States::STARTED
  end
  
  def done?
    state == States::DONE
  end
  
  def failed?
    state == States::FAILED
  end
  
  def state=(value)
    raise ArgumentError, "'#{value}' is not a valid state" unless States::ALL.include?(value)
    super
end
```

With something like:
```ruby
class SomeModel
  attribute :state, String
  enum :state, [:started, :done, :failed]
end
```

Would you be open to an ActiveModel::EnumType?
So we can do something like:
```ruby
class SomeModel
  attribute :state, :enum, [:started, :done, :failed]
end
```

rafaelfranca commented:
> My main reason for an ActiveModel::Enum is to replace code like the following:

That is exactly my problem with `Enum`. Look how many options he have to disable those methods. It is not like people always need them, and people sometimes don't want them. So not sure it is worth to add a a DSL to generate those methods given them aren't hard to implement.

A type is an interesting idea, but not sure it solve the problem you are trying to solve. Types don't generate methods. Neither do validation, they usually make invalid values `nil`.
